dbt/adapters/record.py

```diff
@@ -1,67 +1,206 @@
 import dataclasses
-from io import StringIO
-import json
-import re
-from typing import Any, Optional, Mapping
+import datetime
+from typing import Any, Dict, Optional, Mapping, List, Union, Iterable
 
-from agate import Table
+from dbt.adapters.contracts.connection import Connection
 
-from dbt_common.events.contextvars import get_node_info
-from dbt_common.record import Record, Recorder
+from dbt_common.record import Record, Recorder, record_function
 
-from dbt.adapters.contracts.connection import AdapterResponse
+
+class RecordReplayHandle:
+    def __init__(self, native_handle: Any, connection: Connection) -> None:
+        self.native_handle = native_handle
+        self.connection = connection
+
+    def cursor(self):
+        # The native handle could be None if we are in replay mode, because no
+        # actual database access should be performed in that mode.
+        cursor = None if self.native_handle is None else self.native_handle.cursor()
+        return RecordReplayCursor(cursor, self.connection)
+
+
+@dataclasses.dataclass
+class CursorExecuteParams:
+    connection_name: str
+    operation: str
+    parameters: Union[Iterable[Any], Mapping[str, Any]]
+
+
+@Recorder.register_record_type
+class CursorExecuteRecord(Record):
```

Comment:
one other thought: not all adapters actually use 'cursors' and I _think_ what we're actually talking about 'connections'. It may just make more sense to drop the `Cursor*` prefix entirely